src/pip/_internal/req/req_install.py

```diff
@@ -486,16 +503,70 @@ def load_pyproject_toml(self) -> None:
             backend_path=backend_path,
         )
 
-    def _generate_metadata(self) -> str:
+    def _generate_editable_metadata(self) -> str:
         """Invokes metadata generator functions, with the required arguments."""
-        if not self.use_pep517:
-            assert self.unpacked_source_directory
+        if self.use_pep517:
+            assert self.pep517_backend is not None
+            try:
+                metadata_directory = generate_editable_metadata(
+                    build_env=self.build_env,
+                    backend=self.pep517_backend,
+                )
+            except HookMissing as e:
+                self.supports_pep660 = False
+                if not os.path.exists(self.setup_py_path) and not os.path.exists(
+                    self.setup_cfg_path
+                ):
+                    raise InstallationError(
+                        f"Project {self} has a 'pyproject.toml' and its build "
+                        f"backend is missing the {e} hook. Since it does not "
+                        f"have a 'setup.py' nor a 'setup.cfg', "
+                        f"it cannot be installed in editable mode. "
+                        f"Consider using a build backend that supports PEP 660."
+                    )
+                # At this point we have determined that the build_editable hook
+                # is missing, and there is a setup.py or setup.cfg
+                # so we fallback to the legacy metadata generation
+            else:
+                self.supports_pep660 = True
+                return metadata_directory
+        else:
+            if not os.path.exists(self.setup_py_path) and not os.path.exists(
+                self.setup_cfg_path
+            ):
+                raise InstallationError(
+                    f"File 'setup.py' or 'setup.cfg' not found "
+                    f"for legacy project {self}. "
+                    f"It cannot be installed in editable mode."
+                )
```

Comment:
```suggestion
        elif not (
            os.path.exists(self.setup_py_path) or os.path.exists(self.setup_cfg_path)
        ):
            raise InstallationError(
                f"File 'setup.py' or 'setup.cfg' not found "
                f"for legacy project {self}. "
                f"It cannot be installed in editable mode."
            )
```

Ooh, we can simplify this!